### PR TITLE
Generate warning when overflow happen during const expr cast

### PR DESCRIPTION
**Pull request by vcsharma**
Generate warning when overflow happen during const expr cast

### CODE REVIEW

File: tools/clang/include/clang/Basic/DiagnosticASTKinds.td [hunk 1]
```diff
@@ -263,4 +263,8 @@ def err_odr_non_type_parameter_type_inconsistent : Error<
   "non-type template parameter declared with incompatible types in different "
   "translation units (%0 vs. %1)">;
 def err_unsupported_ast_node: Error<"cannot import unsupported AST node %0">;
+
+// HLSL changes begin
+def warn_hlsl_constexpr_overflow : Warning<"overflow in the expression">;
```
Comment (tex3d) [inline tools/clang/include/clang/Basic/DiagnosticASTKinds.td:268]: We should get in the habit of using diagnostic groups for warnings (InGroup<>, see DiagnosticGroups.td, or define one-off inline with DiagGroup<"some-group-name">).  See `warn_integer_constant_overflow` for example.  By the way, that warning could be used here, couldn't it?  There's a related one for float, but it's just a note: `note_constexpr_float_arithmetic`

Comment (pow2clk) [reply tools/clang/include/clang/Basic/DiagnosticASTKinds.td:268]: I was wondering if there was a reason not to to convert that note in HandleOverflow to a warning and produce warnings for all these cases. There are a couple places besides the ones enhanced here, but not many.

Wouldn't mind considering that for a followup either.

Comment (vcsharma) [reply tools/clang/include/clang/Basic/DiagnosticASTKinds.td:268]: I think we can move this warning inside HandleOverflow and change the note itself to warn.

Comment (vcsharma) [reply tools/clang/include/clang/Basic/DiagnosticASTKinds.td]: I wanted to use `warn_integer_constant_overflow` except that its wording "..result is %0 with type %1.." would require printing converted result which is not passed to HandleOverflow. So I decided to keep the hlsl version of warning which is informative but does not require passing the result. also added it to a diag group..

**Pull request merged**